### PR TITLE
ignore last void parameter

### LINKED ISSUE #4260: Syntax suggestion: Ignore last void parameters
``` typescript

class X<T>
{
    f(t: T) { return { a: t }; }
}


var x: X<number>;
var t1 = x.f(5);
t1.a = 5; // Should not error: t1 should have type {a: number}, instead has type {a: T}

```

``` typescript
var x2: X<void>; 
var t2a = x.f(void 0); // It works but void 0 it's unnecessary
var t2b = x.f(); // SUGGESTION: ignore syntax check of last argument if it's type is void
t2b.a = void 0; // Should not error: t2 should have type {a: void}, instead has type {a: T}

```

test case: tests/cases/compiler/genericObjectLitReturnType.ts

maybe the solution is the same for #3356

### PR DESCRIPTION
**Pull request by Kingwl**
Close #20570 
Fixes #4260

### CODE REVIEW

File: tests/baselines/reference/arrayFrom.types [hunk 1]
```diff
@@ -37,7 +37,7 @@ const inputALike: ArrayLike<A> = { length: 0 };
 const inputARand = getEither(inputA, inputALike);
 >inputARand : ArrayLike<A> | Iterable<A>
 >getEither(inputA, inputALike) : ArrayLike<A> | Iterable<A>
->getEither : <T>(in1: Iterable<T>, in2: ArrayLike<T>) => Iterable<T> | ArrayLike<T>
```
Comment (Kingwl) [inline tests/baselines/reference/arrayFrom.types]: i don't know why they swapped

**Comment by Kingwl** [comment]
@mhegazy @DanielRosenwasser 
review need please😢

**Comment by Kingwl** [comment]
friendly ping 😂

**Comment by whtsky** [comment]
this pr is very helpful to me

**Comment by mhegazy** [comment]
I do not think `void` is the right type. `void` means the absence of a return type. it should only be used as a return type for a function. 
Now for the original issue. i think we need to have a holistic approach to this issue. there are a few problems here that need to be addressed.. first `undefined` and missing distinction (see https://github.com/Microsoft/TypeScript/issues/13195). second, what does that mean in the context of tuple types (since ideally all parameter lists can be transformed to a tuple tuple, and vice versa), and any thing we do in this space needs to either enable or at least not conflict with variadic types (see https://github.com/Microsoft/TypeScript/issues/5453)

**Pull request closed without merging**